### PR TITLE
Dogfood new_uninit and maybe_uninit_slice in rustc_arena

### PR DESCRIPTION
**Pull request by est31**
Dogfoods a few cool `MaybeUninit` related features in the compiler's rustc_arena crate.

Split off from #76821

r? @oli-obk

### CODE REVIEW

File: compiler/rustc_arena/src/lib.rs [hunk 1]
```diff
@@ -69,30 +66,25 @@ impl<T> TypedArenaChunk<T> {
         // The branch on needs_drop() is an -O1 performance optimization.
         // Without the branch, dropping TypedArena<u8> takes linear time.
         if mem::needs_drop::<T>() {
-            let mut start = self.start();
-            // Destroy all allocated objects.
-            for _ in 0..len {
-                ptr::drop_in_place(start);
-                start = start.offset(1);
-            }
+            ptr::drop_in_place(ptr::slice_from_raw_parts_mut(self.start(), len));
```
Comment (oli-obk) [inline compiler/rustc_arena/src/lib.rs:69]: What do you think about 

```suggestion
            ptr::drop_in_place(MaybeUninit::slice_get_mut(&mut self.storage[..len]));
```

**Comment by est31** [comment]
Does this need a perf run? I see some regressions in the microbenchmarks, but being microbenchmarks they might lie. Did `./x.py bench --stage 0 compiler/rustc_arena`.

Output with base commit from bors:

```
test tests::bench_copy              ... bench:           5 ns/iter (+/- 0)
test tests::bench_copy_nonarena     ... bench:           0 ns/iter (+/- 0)
test tests::bench_noncopy           ... bench:          64 ns/iter (+/- 1)
test tests::bench_noncopy_nonarena  ... bench:          34 ns/iter (+/- 1)
test tests::bench_typed_arena_clear ... bench:           5 ns/iter (+/- 0)
```

Output with only the first commit:

```
test tests::bench_copy              ... bench:           5 ns/iter (+/- 0)
test tests::bench_copy_nonarena     ... bench:           0 ns/iter (+/- 0)
test tests::bench_noncopy           ... bench:          65 ns/iter (+/- 1)
test tests::bench_noncopy_nonarena  ... bench:          36 ns/iter (+/- 1)
test tests::bench_typed_arena_clear ... bench:           7 ns/iter (+/- 0)
```

Output with both commits:

```
test tests::bench_copy              ... bench:           5 ns/iter (+/- 0)
test tests::bench_copy_nonarena     ... bench:           0 ns/iter (+/- 0)
test tests::bench_noncopy           ... bench:          63 ns/iter (+/- 1)
test tests::bench_noncopy_nonarena  ... bench:          33 ns/iter (+/- 1)
test tests::bench_typed_arena_clear ... bench:           6 ns/iter (+/- 0
```

There is tons of noise in the benchmarks but I made multiple runs each time and there is a consistent difference. There might be a regression in `bench_typed_arena_clear`. Again, it's a microbenchmark and the usage pattern of the arena in the compiler might not reflect it at all.

If I allocate 100 items and do the same clearance I can't find a regression, in fact it's a bit faster now. Benchmark:

```Rust
#[bench]
pub fn bench_typed_arena_clear_100(b: &mut Bencher) {
    let mut arena = TypedArena::default();
    b.iter(|| {
        for _ in 0..100 {
            arena.alloc(Point { x: 1, y: 2, z: 3 });
        }
        arena.clear();
    })
}
```

Before this PR:

```
test tests::bench_typed_arena_clear_100 ... bench:         177 ns/iter (+/- 5)
```

After this PR:

```
test tests::bench_typed_arena_clear_100 ... bench:         173 ns/iter (+/- 12)
```

Again, lot's of noise.

I've added the benchmark to this PR. I've also amended the first commit to do a `RawVec → Box` change in one of the comments. re-r? @oli-obk

**Comment by oli-obk** [comment]
@bors r+ rollup=never

I don't think this will have an effect on perfbot, but let's not roll this up to make sure

**Pull request merged**